File: cpp/Add_Two_Numbers.cc
/*
You are given two linked lists representing two non-negative numbers. The digits are stored in reverse order and each of their nodes contain a single digit. Add the two numbers and return it as a linked list.

Input: (2 -> 4 -> 3) + (5 -> 6 -> 4)
Output: 7 -> 0 -> 8
*/

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution 
{
public:
    ListNode *addTwoNumbers(ListNode *l1, ListNode *l2) 
    {
        ListNode* head = NULL;
        ListNode** pCur = &head;
        
        int carry = 0;
        while (l1 != NULL && l2 != NULL)
        {
            ListNode* tmp = new ListNode((l1->val + l2->val + carry) % 10);
            *pCur = tmp;
            pCur = &((*pCur)->next);
            carry = (l1->val + l2->val + carry) / 10;
            
            l1 = l1->next;
            l2 = l2->next;
        }
        
        while (l1 != NULL)
        {
            ListNode* tmp = new ListNode((l1->val + carry) % 10);
            *pCur = tmp;
            pCur = &((*pCur)->next);
            carry = (l1->val + carry) / 10;
            
            l1 = l1->next;
        }
        
        while (l2 != NULL)
        {
            ListNode* tmp = new ListNode((l2->val + carry) % 10);
            *pCur = tmp;
            pCur = &((*pCur)->next);
            carry = (l2->val + carry) / 10;
            
            l2 = l2->next;
        }
        
        if (carry > 0)
        {
            ListNode* tmp = new ListNode(carry);
            *pCur = tmp;
        }
        
        return head;
    }
/*
    // 增加一个dummy head。优化处理流程
    ListNode *addTwoNumbers(ListNode *l1, ListNode *l2)
    {
        ListNode dummy(0);
        ListNode* pCur = &dummy;
        
        int carry = 0;
        while (l1 != NULL || l2 != NULL)
        {
            ListNode* tmp = new ListNode((getValue(l1) + getValue(l2) + carry) % 10);
            pCur->next = tmp;
            pCur = pCur->next;
            
            carry = (getValue(l1) + getValue(l2) + carry) / 10;
            if (l1 != NULL)
                l1 = l1->next;
            if (l2 != NULL)
                l2 = l2->next;
        }
        
        if (carry > 0)
        {
            ListNode* tmp = new ListNode(carry);
            pCur->next = tmp;
        }
        
        return dummy.next;
    }
private:
    int getValue(ListNode* &l)
    {
        int ret = 0;
        if (l != NULL)
        {
            ret = l->val;
        }
        return ret;
    }
*/
};

File: go/add-two-numbers/add-two-numbers.go
package add_two_numbers

import "../util"

/*
2. 两数相加

给定两个非空链表来表示两个非负整数。位数按照逆序方式存储，它们的每个节点只存储单个数字。将两数相加返回一个新的链表。

你可以假设除了数字 0 之外，这两个数字都不会以零开头。

示例：

输入：(2 -> 4 -> 3) + (5 -> 6 -> 4)
输出：7 -> 0 -> 8
原因：342 + 465 = 807
*/

type ListNode = util.ListNode

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	dummy := &ListNode{-1, nil}
	cur := dummy

	carry := 0
	for l1 != nil || l2 != nil || carry > 0 {
		num := carry

		if l1 != nil {
			num += l1.Val
			l1 = l1.Next
		}

		if l2 != nil {
			num += l2.Val
			l2 = l2.Next
		}

		carry = num / 10

		cur.Next = &ListNode{Val: num % 10}
		cur = cur.Next
	}

	return dummy.Next
}
